gpu_direct/gpu.c

```diff
@@ -0,0 +1,326 @@
+
+    // Get devices for driver
+    uint32_t devicesCount = 0;
+    ZE_CHECK_ERROR(zeDeviceGet(driver, &devicesCount, NULL));
+    ze_device_handle_t* devices = calloc(devicesCount, sizeof(ze_device_handle_t));
+    if (!devices) {
+      fprintf(stderr, "Memory allocation for devices failed\n");
+      free(drivers);
+      return -1;
+    }
+    ZE_CHECK_ERROR(zeDeviceGet(driver, &devicesCount, devices));
+    for (int j = 0; j < devicesCount; j++) {
+      ze_device_handle_t device = devices[j];
+      // Get properties of the device
+      ze_device_properties_t deviceProperties;
+      ZE_CHECK_ERROR(zeDeviceGetProperties(device, &deviceProperties));
+      printf("Driver: %d: Device: %d: Name: %s, Type: %d, VendorID: %x, DeviceID: %d\n",
+             i, j, deviceProperties.name, deviceProperties.type,
+             deviceProperties.vendorId, deviceProperties.deviceId);
+    }
+    free(devices);
+  }
+
+  free(drivers);
+  return 0;
+}
+
+/**
+ * @brief Init GPU device. Driver index and device index is required.
+ * To get indexes, use GetGPUDriversAndDevices().
+ * After finishing working with the context calling FreeGpuContext(ctx) is required to
+ * free the resources.
+ *
+ * @param ctx - [in/out] GPU context to init device, context will be updated during
+ * initialization.
+ * @param driverIndex - [in] driver index.
+ * @param deviceIndex - [in] device index.
+ * @return int. 0 if successfull, if error occured -1.
+ */
+int init_gpu_device(GpuContext* ctx, unsigned driverIndex, unsigned deviceIndex) {
+  if (ctx->initialized) {
+    fprintf(stderr, "Context is already initialized\n");
+    return -1;
+  }
+
+  // Init level-zero lib
+  INIT_CHECK_ERROR(init_level_zero_lib());
+
+  // Get drivers count
+  ZE_CHECK_ERROR(zeDriverGet(&ctx->driverCount, NULL));
+  if (driverIndex >= ctx->driverCount) {
+    fprintf(stderr, "Init error: provided driver index is out of range\n");
+    return -1;
+  }
+
+  // Allocate and retrieve driver handlers
+  ctx->drivers = calloc(ctx->driverCount, sizeof(ze_driver_handle_t));
+  if (ctx->drivers == NULL) {
+    fprintf(stderr, "Can't allocation memory for drivers handlers\n");
+    return -1;
+  }
+  ZE_CHECK_ERROR(zeDriverGet(&ctx->driverCount, ctx->drivers));
+  ctx->currentDriverIndex = driverIndex;
+  ctx->driverHandle = ctx->drivers[driverIndex];
+
+  // Init context
+  ze_context_desc_t ctxDesc = {.stype = ZE_STRUCTURE_TYPE_CONTEXT_DESC};
+  ZE_CHECK_ERROR(zeContextCreate(ctx->driverHandle, &ctxDesc, &ctx->deviceContext));
+
+  // Get decive count
+  ZE_CHECK_ERROR(zeDeviceGet(ctx->driverHandle, &ctx->deviceCount, NULL));
+  printf("Number of devices: %d\n", ctx->deviceCount);
+  if (deviceIndex >= ctx->deviceCount) {
+    fprintf(stderr, "Init error: provided device index is out of range\n");
+    return -1;
+  }
+
+  // Allocate and retrieve device handlers
+  ctx->devices = calloc(ctx->deviceCount, sizeof(ze_device_handle_t));
+  if (ctx->devices == NULL) {
```

Comment:
```suggestion
  if (!ctx->devices) {
```

Reply:
Hi with this i don't think i agree

Although similar the ctx->devices == NULL  is a lil bit more readily apparent then ! 
if i read the code with pointers i much prefer the == NULL

Reply:
Agree. But there is a common Linux community practice to check that with `if (!ptr)`. For example, you'll hardly find `if (ptr == NULL)` or `if (ptr != NULL)` in the code of Linux drivers.